### ISSUE TITLE
Implicit-AD gradient magnitude is ~15x too large on a D=3 iPEPS state (direction valid, scale is not); no line search can descend it

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#841]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

`ctm_energy_implicit` returns a gradient whose **direction is a valid descent direction but whose magnitude is overstated by ~15×** at a D=3 iPEPS state. The discrepancy is measured against a finite difference of the *same function whose gradient was taken*, so it is self-consistent and does not depend on any reference implementation.

Practical effect: no line search can work. `optimize_gs_ad` opens at `alpha0 = min(1.0, 0.1*param_norm/dir_norm)`, which on this state is ~26× past the largest useful displacement, and both line-search backends fail — Hager-Zhang in a ~7-step limit cycle, Armijo in a hard stall at `dE = 0.000e+00`. Neither reports anything wrong: the adjoint emits **zero** non-convergence warnings.

## Measurement

State: D=3, `E = -0.660153228693`, `chi = 27`. Gradient via `jax.value_and_grad` of

```python
def energy(arr):
    A = wrap(arr / jnp.linalg.norm(arr))          # DenseTensor, trivial U(1) charges
    return ctm_energy_implicit(
        {(0, 0): A}, SINGLE_SITE_NEIGHBORS, GATE,
        chi=27, max_iter=100, conv_tol=1e-10,
        projector_method="svd", conv_method="elementwise", recipe="2x2",
        forward_gauge="phase", min_iter=10,
        gmres_tol=1e-6, gmres_maxiter=500, gmres_restart=50,
        adjoint_method="fixed_point",
    )                                              # env_init=None -> cold, pure
```

`env_init` is left `None` so the function is pure; note this deliberately does **not** reuse `optimize_gs_ad`'s `loss_fn`, which closes over a mutable `env_cache` (see the secondary finding below).

```
         t      E(A - t*ghat)             dE     predicted -t|g|     ratio
   1.0e-01   -0.653188810037      +6.964e-03       -5.492e-03      -1.268
   3.0e-02   -0.659574647935      +5.786e-04       -1.647e-03      -0.351
   1.0e-02   -0.660112560146      +4.067e-05       -5.492e-04      -0.074
   3.0e-03   -0.660157117845      -3.889e-06       -1.647e-04      +0.024
   1.0e-03   -0.660156059178      -2.830e-06       -5.492e-05      +0.052
   3.0e-04   -0.660154238955      -1.010e-06       -1.647e-05      +0.061
   1.0e-04   -0.660153580792      -3.521e-07       -5.492e-06      +0.064
```

Central-difference directional derivative along `+ghat`, which should equal `|g|`:

```
conv_method="elementwise"   |g| = 5.491625e-02
  t=1e-03   slope_fd = +3.597771e-03   ratio = 0.0655

conv_method="sv"            |g| = 5.493990e-02
  t=1e-03   slope_fd = +3.588209e-03   ratio = 0.0653
  t=1e-04   slope_fd = +3.588157e-03   ratio = 0.0653
```

`slope_fd` is stable to five digits across two step sizes and across both convergence criteria. **The true directional derivative is 6.5% of the reported gradient norm — a factor of 15.3.**

The energy does decrease along `-g`, but only for displacements `t <~ 3e-3`; by `t = 1e-2` it is rising. `dE` at the smallest step is `3.5e-07`, still ~3500× above the `~1e-10` reproducibility floor of the energy, so this is not evaluation noise.

Sanity check that passes: `<ghat, A>/|A|` is `1.27e-16` (elementwise) and `3.08e-17` (sv), i.e. the gradient is orthogonal to `A` exactly as scale-invariance of the energy requires.

## Ruled out by measurement

- **CTM convergence criterion.** `conv_method="sv"` gives `0.0653` against `elementwise`'s `0.0655`. Identical. So the concern in #780 — that the default `elementwise` exits via `plateau_patience` rather than at the fixed point, violating implicit AD's premise — does **not** explain this. (Useful negative datum for #780.)
- **Adjoint non-convergence.** Zero `adjoint solve did not converge` warnings for the entire run at `chi=27, adjoint_tikhonov=1e-6` (i.e. unbiased, no Tikhonov damping in play).
- **Corner degeneracy / rank deficiency.** At `chi=27` the corner `C1` is rank 27/27 with minimum adjacent normalized singular-value gap `1.14e-06` and `s_min/s_max = 3.06e-05` — better conditioned than a D=2 state on which implicit AD works correctly (`8.99e-08` and `3.66e-06`). *Caveat: this spectrum was measured on the state two optimizer steps upstream of the frozen point, not the frozen point itself.*

## Live hypothesis: gauge contamination

The iPEPS energy is invariant under virtual-bond gauge transformations, so gauge directions are a null space of the true gradient. Writing `g = g_phys + g_gauge`, the measured slope is `|g_phys|²/|g|`, giving

```
|g_phys| = 1.40e-02      |g| = 5.49e-02      -> ~74% of the reported norm is non-descending
```

Untested knob: `CTMConfig.forward_gauge` (`"phase"` default, `"sigma"` 1-site only, `"qr"`, `"none"`). If a different forward gauge moves `ratio` toward 1, that localizes it.

## Secondary finding: warm-cache gradient differs by 41%

`optimize_gs_ad` builds its loss with `make_ctm_energy_fn(..., env_cache=_env_cache, ...)`, warm-starting the CTM from a cache mutated between steps, so `loss_fn` is not a pure function of the parameters. At the identical state:

```
optimizer (warm-started env_cache)   |g| = 7.770e-02
cold ctm_energy_implicit             |g| = 5.492e-02      -> 41% apart
```

Forward energies agree to twelve digits; only the backward differs. Possibly the same root cause, possibly separate, but it means the optimizer descends along a direction that differs measurably from the gradient of the function it evaluates.

## Why this is not an existing issue

- **#824 / #827** (`test_fixed_point_matches_gmres_gradient`) concern a gradient ~3% unconverged — three orders of magnitude milder.
- **#785** is about undetectable inaccurate *root-implicit* gradients; this ran the default `ctm_ad_mode=None`, not a root-implicit mode.
- **#687** is block-sparse SVD backward; this is the dense path.
- **#780** is addressed above and is refuted as the cause.

## Reproducing

The state is 2 `optimize_gs_ad` steps (`chi=27`, L-BFGS) from a zero-padded D=2 optimum plus `eps=0.1` noise, and reproduces bit-exactly (`-0.660153228693`). Happy to attach the `.npy` and the two probe scripts if useful.

**Environment:** tenax `c476422`, jax/jaxlib 0.10.1, A100 (CUDA).

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Full measurement campaign on current `main` (9bf90f8), CPU x64. **Reproduced** — and the live hypothesis in the issue body is refuted.

## Reproduction

The literal 2-step construction from the issue gives ratio **0.9991** (phenomenon absent); proximity to optimality is what matters, not the padding recipe. After 40 more L-BFGS steps at χ=27 (E=−0.665240, |g|=3.7e−2) the ratio is **0.1310** (slope_fd stable at t=1e−3/1e−4) — the issue's class (their 0.065 at their exact state). An SU-relaxed D=3 state gives 0.9995: benign away from optimality.

## Gauge contamination is NOT the cause

Built the gauge-orbit tangent space at A (rank 16/18, invariance verified through the CTM energy to O(ε²)). The orbit component of g is real, pure contamination (FD along it: true slope ~−5e−9 vs AD's claimed +5.7e−3) — but it holds only **~27% of the error norm** (frac² = 6.2% of |g|², not the predicted ~74%). The excess is mostly *outside* the gauge orbit. The `forward_gauge` knob is not the fix: `none` is worse (0.238), `qr` does not exist (the issue's knob list was wrong), and `sigma` converges to an outright wrong environment on the 2×2 recipe (E=−0.6667 vs −0.6563, |g|=1.3e2) — which matches #798's predicted mechanism.

## Root cause: the implicit-AD premise is false at D=3

The phase-gauged 2×2 forward **never reaches an element-wise fixed point on any D=3 state tested** — persistent O(0.5) sup-norm steps for 200+ sweeps at χ=24/27/32 — while the gauge-invariant content fully converges (2-site RDM Δ→1e−15 by sweep ~40, E stable to 1e−12). The forward exits at `max_iter` with `converged=False` and the flag is silently discarded (`_ctm_energy_ad.py:834` returns only `(envs, chi_post)`). The backward then linearizes `phase_fix(step(·))` around a non-stationary point, with the projector response additionally `stop_gradient`-ed. The residual per-χ-index gauge freedom (Z₂^χ signs + O(k) rotations inside near-degenerate corner multiplets) is what prevents element-wise stationarity — the corner min-gap 1.14e−6 the issue measured and ruled out is exactly the fingerprint.

**Control that closes the loop:** `recipe="1x1"` — whose forward genuinely converges (to its known rank-collapsed fixed point) and whose projectors are differentiable — gives ratio **1.0000** on the same state. The implicit machinery is exact when its premise holds.

Also refuted along the way: solver/tolerance (gmres_tol 1e−6→1e−10 moves g by 6.7e−7, ratio unchanged; adjoint residual 2.4e−11), and the projector `stop_gradient` as *sole* cause (a differentiable-projector patch with Lorentzian ε=1e−12 stays finite but ratio 0.090; that defect is real but secondary — filed separately).

## In flight

A fix branch is being built: (1) an honesty guard — measure the literal stationarity residual ‖gauge_fix(step(env*)) − env*‖ after the forward and warn when the linearization premise fails (today's silent 7–15× wrong gradient becomes a diagnosable event); (2) repair of the sigma gauge on the 2×2 recipe (#798), gated on whether it restores element-wise stationarity and drives the ratio to ~1 on the saved reproducing state. The projector `stop_gradient` half is filed as its own issue.

Caveats: CPU-only (issue measured on A100 — GPU kernels unmeasured); GATE inferred as the sublattice-rotated Heisenberg gate (supported by the D=2 optimum matching to 2e−5); the non-stationarity and projector halves are entangled — the 1x1 control fixes both at once.
